include/Graph/Graph.hpp

```diff
@@ -805,7 +868,32 @@ bool Graph<T>::findEdge(shared<const Node<T>> v1, shared<const Node<T>> v2,
                         unsigned long long &id) const {
   // This could be made faster by looking for the edge hash, assuming we hash
   // based on node data, instead of a unique integer
-  for (auto e : this->edgeSet) {
+  if(cachedAdjMatrix.get() != NULL){
+    /* Searching for the edge using cached adjacency matrix */
+
+    for(auto elem : (*cachedAdjMatrix)[v1]){
+      if(elem.first == v2){
+        id = elem.second.get()->getId();
+        return true;
+      }
+    }
+  }
+  else{
```

Comment:
Maybe it could be possible to add tests for the condition of this else statement? codecov is giving some warnings

Reply:
You enter in the else statement when the cached matrix is `NULL`, and if I'm not mistaken this happens when you construct the graph with the default constructor and then add the edges. It would suffice to test this condition

Reply:
yes. I guess the else part will only be hit for empty graph and the subsequent for loop won't be covered for an empty graph. Also, in the method 'getAdjMatrix()' the partially populated matrix gets returned if the edge is of type 'Edge<T>'(line number 1497). This can probably cause issues. Btw, what does an edge of type 'Edge<T>' signify?

Reply:
The base class `Edge` represents an abstract edge, and defines the fundamental methods that every edge needs. Then this is derived in order to distinguish directed/undirected and weighted/unweighted edges.

Reply:
okay then Edge class can be an abstract class? So that it doesn't allow to add edges of type 'Edge' in the graph?. Right now we can add edge of type 'Edge' in the edgeset but it will not get added to the adj matrix.

Reply:
Looking at the Edge class structure now, it's clear we have some overlap between inheritance and composition. It provides a bit of an odd scenario where Edge and UndirectedEdge are basically the same class.

That doesn't necessarily need to be resolved in this PR (and shouldn't be so that reviewing is easier), however I think it would bring clarity to force a direction here:
1. We make Edge abstract and duplicate some implementation between Undirected and Directed
2. We remove UndirectedEdge and Edge implicitly is undirected (which we document)

Personally I like (2.) a bit more since it reduces the amount of code, at the cost of some name verbosity (although in-reality we could rename Edge -> UndirectedEdge).


All this being said, I don't think the above should be a blocker to this PR. We can simply treat an `Edge` class like an UndirectedEdge for now.

Reply:
@ZigRazor ☝️ Let's make sure to follow-up on this

Reply:
Ok the new tests look good. Thank you :)